File: Python/data_structure/weighted_union_find.py
# 重み付きUnion-Find
class WeightedUnionFind:
	def __init__(self, n: int) -> None:
		self.n = n
		self.par = list(range(n))
		self.rank = [0] * n
		self.weight = [0] * n

	def find(self, x: int) -> int:
		if self.par[x] == x:
			return x
		else:
			y = self.find(self.par[x])
			self.weight[x] += self.weight[self.par[x]]
			self.par[x] = y
			return y

	def unite(self, x: int, y: int, w: int) -> None:
		p, q = self.find(x), self.find(y)
		if self.rank[p] < self.rank[q]:
			self.par[p] = q
			self.weight[p] = w - self.weight[x] + self.weight[y]
		else:
			self.par[q] = p
			self.weight[q] = -w - self.weight[y] + self.weight[x]
			if self.rank[p] == self.rank[q]:
				self.rank[p] += 1

	def same(self, x: int, y: int) -> bool:
		return self.find(x) == self.find(y)

	def diff(self, x: int, y: int) -> int:
		return self.weight[x] - self.weight[y]

File: Cpp/data_structure/weighted_union_find.cpp
#include <vector>
#include <cassert>

template <typename T>
struct weighted_union_find {
	public:
		weighted_union_find(int size) {
			n = size;
			parent.resize(n);
			rank.resize(n);
			weight.resize(n);
			for (int i = 0; i < n; i++) parent[i] = i;
		}

		int find(int x) {
			assert(0 <= x && x < n);
			if (parent[x] == x) return x;
			else {
				int y = find(parent[x]);
				weight[x] += weight[parent[x]];
				parent[x] = y;
				return y;
			}
		}

		void unite(int x, int y, T w) {
			assert(0 <= x && x < n);
			assert(0 <= y && y < n);
			int p = find(x), q = find(y);
			if (rank[p] < rank[q]) {
				parent[p] = q;
				weight[p] = w - weight[x] + weight[y];
			} else {
				parent[q] = p;
				weight[q] = -w - weight[y] + weight[x];
				if (rank[p] == rank[q]) rank[p]++;
			}
		}

		bool same(int x, int y) {
			assert(0 <= x && x < n);
			assert(0 <= y && y < n);
			return find(x) == find(y);
		}

		T difference(int x, int y) {
			return weight[x] - weight[y];
		}
	private:
		int n;
		std::vector<int> parent, rank;
		std::vector<T> weight;
};